yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+        if fname is None:
+            raise ValueError("A filename must be specified to load the octtree!")
+        # clear any current tree we have loaded
+        self.reset()
+
+        cdef Node temp
+        cdef np.int64_t num_octs
+        with open(fname,'rb') as f:
+            (self._num_particles, num_octs, self._n_ref,
+             self.over_refine_factor, self.density_factor,
+             self._data_version) = \
+                struct.unpack('2Q3iq', f.read(40))
+            self._idx = \
+                np.asarray(struct.unpack('{}Q'.format(self.num_particles),
+                           f.read(8*self.num_particles)), dtype=np.int64)
+            reserve(&self.nodes, num_octs+1)
+            for i in range(num_octs):
+                (temp.left_edge[0], temp.left_edge[1], temp.left_edge[2],
+                 temp.right_edge[0], temp.right_edge[1], temp.right_edge[2],
+                 temp.start, temp.end, temp.parent, temp.children, temp.leaf,
+                 temp.node_id, temp.leaf_id, temp.depth) = \
+                struct.unpack('6d4Q?2Qi', f.read(108))
+
+                self.nodes.push_back(temp)
+
+        for i in range(3):
+            self._left_edge[i] = self.nodes[0].left_edge[i]
+            self._right_edge[i] = self.nodes[0].right_edge[i]
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int split_helper(int start, int end, int num_children, int max_splits,
+                       np.int64_t * split_arr, np.float64_t * pos,
+                       np.int64_t * idx, np.float64_t * left,
+                       np.float64_t * right) nogil except -1:
+    cdef double lower = left[0]
+    cdef double upper = right[0]
+
+    split_arr[0] = start
+    split_arr[num_children] = end
+
+    split(0, num_children, max_splits, 0, 0, split_arr, idx, pos, left, right,
+          lower, upper)
+
+    return 0
+
+# This is a utility function which is able to catch errors when we attempt to
+# reserve too much memory
+cdef int reserve(vector[Node] * vec, int amount) except +MemoryError:
+    vec.reserve(amount)
+    return 0
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int seperate(np.float64_t * array, np.int64_t * idx, int offset,  double value,
+                  np.int64_t start, np.int64_t end) nogil except -1:
+    cdef np.int64_t index
+    cdef np.int64_t idx_index = start // 3, idx_split = idx_index
+    cdef np.int64_t split = start
+
+    for index in range(start, end, 3):
+        idx_index += 1
+        if array[index + offset] < value:
+            idx[idx_split], idx[idx_index] = idx[idx_index], idx[idx_split]
+            array[split], array[index] = array[index], array[split]
+            array[split+1], array[index+1] = array[index+1], array[split+1]
+            array[split+2], array[index+2] = array[index+2], array[split+2]
+            split+=3
+            idx_split+=1
+
+    return split
```

Comment:
this organizes an array so values less than value are on the left, and values greater are on the right and we store the position of this split.